### PR TITLE
Add local LLM support with model setup UI and CUDA CI updates

### PR DESCRIPTION
**Pull request by ItsSimko**
## Summary by Sourcery

Add support for local LLMs with a loading window for model setup, allowing users to choose between local and API-based models. Enhance the installer to include model downloads and update the CI workflow for CUDA support.

New Features:
- Introduce support for local LLMs, allowing users to choose between using a local model or an API for text generation.
- Add a loading window to indicate the model setup process, enhancing user experience during initialization.

Enhancements:
- Refactor the text sending function to support both local models and API-based models, improving flexibility in model usage.
- Implement a mechanism to load or unload models based on user settings, optimizing resource usage.

Build:
- Add a new section in the installer script to download and install the GGUF model, ensuring the necessary model files are available for local LLM usage.

CI:
- Update the CI workflow to include permissions for content writing and install CUDA-enabled llama_cpp for enhanced model performance.

### CODE REVIEW

**Comment by ItsSimko** [comment]
@sourcery-ai review

File: src/FreeScribe.client/UI/SettingsWindow.py [hunk 1]
```diff
@@ -499,3 +511,21 @@ def set_main_window(self, window):
...
+            ModelManager.unload_model()
+            thread = threading.Thread(target=ModelManager.setup_model, args=(self, self.main_window.root))
+            thread.start()
+            print("Reloading model")
+
+        # Load the model if check box is now selected
+        if old_use_local_llm == 0 and new_use_local_llm == 1:
+            thread = threading.Thread(target=ModelManager.setup_model, args=(self, self.main_window.root))
+            thread.start()
+            print("Loading model")
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/SettingsWindow.py:526]: done .

File: src/FreeScribe.client/client.py [hunk 2]
```diff
@@ -603,42 +630,54 @@ def on_proceed():
...
-                    #Make a note from the facts
-                    medical_note = send_text_to_chatgpt(f"{app_settings.AISCRIBE} {list_of_facts} {app_settings.AISCRIBE2}")
-
-                    # If post-processing is enabled check the note over
-                    if app_settings.editable_settings["Use Post-Processing"]:
-                        post_processed_note = send_text_to_chatgpt(f"{app_settings.editable_settings['Post-Processing']}\nFacts:{list_of_facts}\nNotes:{medical_note}")
-                        update_gui_with_response(post_processed_note)
-                    else:
-                        update_gui_with_response(medical_note)
+        def generate_note(formatted_message):
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:633]: done

File: src/FreeScribe.client/client.py [hunk 3]
```diff
@@ -603,42 +630,54 @@ def on_proceed():
...
-                ai_response = send_text_to_chatgpt(edited_text)
-                update_gui_with_response(ai_response)
-        except Exception as e:
-            #Logg
-            #TODO: Implement proper logging to system event logger
-            print(f"An error occurred: {e}")
-            display_text(f"An error occurred: {e}")
-        finally:
-            loading_window.destroy()
+        def check_thread_status(thread, loading_window):
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:674]: the method checks inside the class

File: src/FreeScribe.client/client.py [hunk 4]
```diff
@@ -581,6 +583,31 @@ def send_text_to_chatgpt(edited_text):
     except Exception as e:
         raise e
 
+def send_text_to_localmodel(edited_text):  
+    # Send prompt to local model and get response
+    if ModelManager.local_model is None:
+        ModelManager.setup_model(app_settings=app_settings, root=root)
+
+    while ModelManager.local_model is None:
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:591]: added timeout

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -17,17 +19,23 @@ jobs:
...
+        run: |
+          pip install --index-url https://abetlen.github.io/llama-cpp-python/whl/cu121 --extra-index-url https://pypi.org/simple llama-cpp-python==v0.2.90
       - name: Install requirements
         run: |
           pip install -r client_requirements.txt
+
       - name: Run PyInstaller
         run: |
-          pyinstaller --add-data ".\src\FreeScribe.client\whisper-assets:whisper\assets" --add-data ".\src\FreeScribe.client\markdown:markdown" --add-data ".\src\FreeScribe.client\assets:assets" --name freescribe-client --icon=.\src\FreeScribe.client\assets\logo.ico --noconsole .\src\FreeScribe.client\client.py
+          pyinstaller --additional-hooks-dir=.\scripts\hooks --add-data ".\src\FreeScribe.client\whisper-assets:whisper\assets" --name freescribe-client --icon=.\scripts\assets\logo.ico --noconsole .\src\FreeScribe.client\client.py
```
Comment (pemba1s1) [inline .github/workflows/release.yml:32]:
```suggestion
          pyinstaller --additional-hooks-dir=.\scripts\hooks --add-data ".\src\FreeScribe.client\whisper-assets:whisper\assets" --add-data ".\src\FreeScribe.client\markdown:markdown" --add-data ".\src\FreeScribe.client\assets:assets" --name freescribe-client --icon=.\src\FreeScribe.client\assets\logo.ico --noconsole .\src\FreeScribe.client\client.py
```

File: src/FreeScribe.client/UI/LoadingWindow.py [hunk 6]
```diff
@@ -58,6 +57,9 @@ def __init__(self, parent=None, title="Processing", initial_text="Loading"):
             y = parent.winfo_y() + (parent.winfo_height() - self.popup.winfo_reqheight()) // 2
             self.popup.geometry(f"+{x}+{y}")
             self.popup.transient(parent)
+            
+            #disabled the window
+            parent.wm_attributes('-disabled', True)
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/LoadingWindow.py]: https://github.com/ClinicianFOCUS/FreeScribe/pull/65/commits/1758700547b09ced12dadf0d6ec565342dbc8d0d

File: src/FreeScribe.client/client.py [hunk 7]
```diff
@@ -582,11 +584,78 @@ def send_text_to_chatgpt(edited_text):
...
+    
+
+
+def send_text_to_chatgpt(edited_text):  
+    if app_settings.editable_settings["Use Local LLM"]:
+        return send_text_to_localmodel(edited_text)
+    else:
+        return send_text_to_api(edited_text)
+
+def generate_note(formatted_message):
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:615]: #79 Will need a refactor

File: src/FreeScribe.client/client.py [hunk 8]
```diff
@@ -582,11 +584,78 @@ def send_text_to_chatgpt(edited_text):
...
+                    # If pre-processing is enabled
+                    if app_settings.editable_settings["Use Pre-Processing"]:
+                        #Generate Facts List
+                        list_of_facts = send_text_to_chatgpt(f"{app_settings.editable_settings['Pre-Processing']} {formatted_message}")
+                        
+                        #Make a note from the facts
+                        medical_note = send_text_to_chatgpt(f"{app_settings.AISCRIBE} {list_of_facts} {app_settings.AISCRIBE2}")
+
+                        # If post-processing is enabled check the note over
+                        if app_settings.editable_settings["Use Post-Processing"]:
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:628]: Leave for now as OK, and potentially come back?

**Review by pemba1s1 — CHANGES_REQUESTED** [review]

**Review by yingbull — APPROVED** [review]

**Pull request merged**